test/files/general/pbkvs.tla

```diff
@@ -0,0 +1,1851 @@
+    primary) have a same state.
+********************)
+
+EXTENDS Naturals, Sequences, TLC, FiniteSets
+
+CONSTANT NUM_REPLICAS
+CONSTANT NUM_PUT_CLIENTS
+CONSTANT NUM_GET_CLIENTS
+CONSTANT EXPLORE_FAIL
+
+ASSUME NUM_REPLICAS > 0 /\ NUM_PUT_CLIENTS >= 0 /\ NUM_GET_CLIENTS >= 0
+
+(********************
+
+--mpcal pbkvs {
+
+    define {
+        NUM_NODES == NUM_REPLICAS + NUM_PUT_CLIENTS + NUM_GET_CLIENTS
+
+        CLIENT_SRC == 1
+        PRIMARY_SRC == 2
+        BACKUP_SRC == 3
+
+        GET_REQ == 1
+        GET_RESP == 2
+        PUT_REQ == 3
+        PUT_RESP == 4
+        SYNC_REQ == 5
+        SYNC_RESP == 6
+
+        REQ_INDEX == 1
+        RESP_INDEX == 2
+
+        ACK_MSG_BODY == [content |-> "ack-body"]
+
+        KEY1 == "KEY1"
+        VALUE1 == "VALUE1"
+
+        NODE_SET == 1..NUM_NODES
+        REPLICA_SET == 1..NUM_REPLICAS
+        PUT_CLIENT_SET == (NUM_REPLICAS+1)..(NUM_REPLICAS+NUM_PUT_CLIENTS)
+        GET_CLIENT_SET == (NUM_REPLICAS+NUM_PUT_CLIENTS+1)..(NUM_REPLICAS+NUM_PUT_CLIENTS+NUM_GET_CLIENTS)
+
+        MSG_INDEX_SET == {REQ_INDEX, RESP_INDEX}
+
+        KEY_SET == {KEY1}
+        VALUE_SET == {VALUE1}
+        
+        NULL == 0
+    }
+
+    macro mayFail(selfID, netEnabled) {
+        if (EXPLORE_FAIL) {
+            either { skip; } or {
+                netEnabled[<<selfID, REQ_INDEX>>] := FALSE;
+                netEnabled[<<selfID, RESP_INDEX>>] := FALSE;
+                goto failLabel;
+            };
+        };
+    }
+
+    mapping macro ReliableFIFOLink {
+        read {
+            assert $variable.enabled;
+            await Len($variable.queue) > 0;
+            with (readMsg = Head($variable.queue)) {
+                $variable := [queue |-> Tail($variable.queue), enabled |-> $variable.enabled];
+                yield readMsg;
+            };
+        }
+
+        write {
+            await $variable.enabled;
+            yield [queue |-> Append($variable.queue, $value), enabled |-> $variable.enabled];
+        }
+    }
+
+    mapping macro NetworkToggle {
+        read { yield $variable.enabled; }
+
+        write {
+            yield [queue |-> $variable.queue, enabled |-> $value];
+        }
+    }
+
+    mapping macro PerfectFD {
+        read {
+            yield $variable;
+        }
+
+        write { yield $value; }
+    }
+
+    mapping macro PracticalFD {
+        read {
+            if ($variable = FALSE) { \* process is alive
+                either { yield TRUE; } or { yield FALSE; }; \* no accuracy guarantee
+            } else {
+                yield $variable; \* (eventual) completeness
+            };
+        }
+
+        write { yield $value; }
+    }
+
+    mapping macro FileSystem {
+        read { yield $variable; }
+
+        write { yield $value; }
+    }
+
+    mapping macro LeaderElection {
+        read {
+            if (Cardinality($variable) > 0) {
+                yield CHOOSE x \in $variable: \A r \in $variable: x =< r;
```

Comment:
Did you know that `CHOOSE` does not force TLC to explore all possibilities?

The way to do that is with the `with` statement, which does.

Reply:
The output of this expression should be unique (since it selects the smallest element in the set). So I think it should be no difference if we use `CHOOSE` or `with` in this case.

Reply:
Fair enough... I guess I'm just naturally afraid of `CHOOSE`, because I would be the one to use it at the wrong time. Makes sense though, with a number line that requirement is unique.

Reply:
Do you suggest that it would be better to use `with` in this case?

Reply:
I'll try it to make sure. Also in other places, for example, in line 215 of `pbkvs.tla` I've used `CHOOSE` like this:
```
replica := CHOOSE r \in replicaSet: TRUE;
```

This is intentional and I didn't want to explore all the possibilities.

Reply:
Now I think more, these cases might be worth a short comment, because the intention is not clear otherwise (mistake vs. intentional).

Also curious about your second example... is there a particular reason? Verification performance?

Reply:
Here there was no need to branching, for example, it would be fine if you always take the first element in the set. So for the verification performance, I used `CHOOSE` statement that has no branching behaviour.

Reply:
I added some comments to clarify the `CHOOSE` construct usage. However, I couldn't try `with` construct because of #161. Although, for sure there should be no difference in using `CHOOSE` or `with` in this case.